File: android/src/main/java/com/crazecoder/flutter/mintegral/FlutterMintegralPlugin.java
package com.crazecoder.flutter.mintegral;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.res.Resources;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.crazecoder.flutter.mintegral.activity.SplashAdActivity;
import com.crazecoder.flutter.mintegral.utils.BannerUtil;
import com.crazecoder.flutter.mintegral.utils.InteractiveAdUtil;
import com.crazecoder.flutter.mintegral.utils.InterstitialVideoUtil;
import com.crazecoder.flutter.mintegral.utils.RewardVideoUtil;
import com.mintegral.msdk.MIntegralConstans;
import com.mintegral.msdk.MIntegralSDK;
import com.mintegral.msdk.out.MIntegralSDKFactory;
import com.mintegral.msdk.out.SDKInitStatusListener;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import io.flutter.plugin.common.StandardMessageCodec;

/**
 * FlutterMintegralPlugin
 */
public class FlutterMintegralPlugin implements FlutterPlugin, MethodCallHandler, ActivityAware {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private MethodChannel channel;
    private static Context context;
    private static Activity activity;
    private FlutterPluginBinding flutterPluginBinding;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        this.flutterPluginBinding = flutterPluginBinding;
    }

    // This static function is optional and equivalent to onAttachedToEngine. It supports the old
    // pre-Flutter-1.12 Android projects. You are encouraged to continue supporting
    // plugin registration via this function while apps migrate to use the new Android APIs
    // post-flutter-1.12 via https://flutter.dev/go/android-project-migration.
    //
    // It is encouraged to share logic between onAttachedToEngine and registerWith to keep
    // them functionally equivalent. Only one of onAttachedToEngine or registerWith will be called
    // depending on the user's project. onAttachedToEngine or registerWith must both be defined
    // in the same class.
    public static void registerWith(Registrar registrar) {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), "flutter_mintegral");
        channel.setMethodCallHandler(new FlutterMintegralPlugin());
        activity = registrar.activity();
        context = registrar.context();
    }

    private void checkAndRequestPermission() {
        List<String> lackedPermission = new ArrayList<String>();
        if (!(ContextCompat.checkSelfPermission(context, Manifest.permission.READ_PHONE_STATE) == PackageManager.PERMISSION_GRANTED)) {
            lackedPermission.add(Manifest.permission.READ_PHONE_STATE);
        }
        // 权限都已经有了，那么直接调用SDK
        if (lackedPermission.size() == 0) {
//      fetchSplashAD(this, container, skipView, appId, adId, splashADListener, 0);
        } else {
            // 请求所缺少的权限，在onRequestPermissionsResult中再看是否获得权限，如果获得权限就可以调用SDK，否则不要调用SDK。
            String[] requestPermissions = new String[lackedPermission.size()];
            lackedPermission.toArray(requestPermissions);
            ActivityCompat.requestPermissions(activity, requestPermissions, 1024);
        }
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        if (call.method.equals("initAdSDK")) {
            checkAndRequestPermission();
            if (call.hasArgument("appId") && call.hasArgument("appKey")) {
                String appId = call.argument("appId");
                String appKey = call.argument("appKey");

                MIntegralSDK sdk = MIntegralSDKFactory.getMIntegralSDK();
                Map<String, String> map = sdk.getMTGConfigurationMap(appId, appKey);
                boolean isProtectGDPR = call.argument("isProtectGDPR");
                sdk.setConsentStatus(activity, isProtectGDPR ? MIntegralConstans.IS_SWITCH_OFF : MIntegralConstans.IS_SWITCH_ON);
                sdk.init(map, context);
                boolean isProtectCCPA = call.argument("isProtectCCPA");
                sdk.setDoNotTrackStatus(isProtectCCPA);
            }
            result.success(null);
        } else if (call.method.equals("startSplashAd")) {
            String adUnitId = call.argument("adUnitId").toString();
            String placementId = call.argument("placementId").toString();
            String launchBackgroundIdStr = call.argument("launchBackgroundId");
            int launchBackgroundId;
            if (TextUtils.isEmpty(launchBackgroundIdStr)) {
                launchBackgroundId = -1;
            } else {
                Resources res = activity.getResources();
                launchBackgroundId = res.getIdentifier(launchBackgroundIdStr, "drawable", activity.getPackageName());
            }

            Intent intent = new Intent(activity, SplashAdActivity.class);
            intent.putExtra("adUnitId", adUnitId);
            intent.putExtra("placementId", placementId);
            intent.putExtra("launchBackgroundId", launchBackgroundId);
            activity.startActivity(intent);
            result.success(null);
        } else if (call.method.equals("showBannerAD")) {
            String adUnitId = call.argument("adUnitId").toString();
            String placementId = call.argument("placementId").toString();

            BannerUtil.getInstance().setActivity(activity).createBanner(placementId, adUnitId);
            BannerUtil.getInstance().show(adUnitId);
            result.success(null);
        } else if (call.method.equals("disposeBannerAD")) {
            String adUnitId = call.argument("adUnitId").toString();
            BannerUtil.getInstance().dispose(adUnitId);
            result.success(null);
        } else if (call.method.equals("showInteractiveAD")) {
            String adUnitId = call.argument("adUnitId").toString();
            String placementId = call.argument("placementId").toString();

            InteractiveAdUtil.getInstance().setActivity(activity).show(adUnitId, placementId);
            result.success(null);
        } else if (call.method.equals("showInterstitialVideoAD")) {
            String adUnitId = call.argument("adUnitId").toString();
            String placementId = call.argument("placementId").toString();

            InterstitialVideoUtil.getInstance().setActivity(activity).show(adUnitId, placementId);
            result.success(null);
        } else if (call.method.equals("showRewardVideoAD")) {
            String adUnitId = call.argument("adUnitId").toString();
            String placementId = call.argument("placementId").toString();
            String userId = call.argument("userId");
            String rewardId = call.argument("rewardId");

            RewardVideoUtil.getInstance().setActivity(activity).show(adUnitId, placementId, userId,rewardId);
            result.success(null);
        } else {
            result.notImplemented();
        }

    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }

    @Override
    public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
        activity = binding.getActivity();
        channel = new MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "flutter_mintegral");
        channel.setMethodCallHandler(this);
        context = flutterPluginBinding.getApplicationContext();
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {

    }

    @Override
    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {

    }

    @Override
    public void onDetachedFromActivity() {
        flutterPluginBinding = null;
    }

}

File: android/src/main/kotlin/com/crazecoder/flutter/mintegral/FlutterMintegralPlugin.kt
package com.crazecoder.flutter.mintegral

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.text.TextUtils
import com.crazecoder.flutter.mintegral.activity.SplashAdActivity
import com.crazecoder.flutter.mintegral.manager.BannerAdManager
import com.crazecoder.flutter.mintegral.manager.MIntegralSDKManager
import com.crazecoder.flutter.mintegral.manager.NewInterstitialAdManager
import com.crazecoder.flutter.mintegral.manager.RewardVideoAdManager
import com.crazecoder.flutter.mintegral.factory.BannerViewFactory
import com.crazecoder.flutter.mintegral.factory.NativeAdViewFactory
import com.crazecoder.flutter.mintegral.factory.NativeAdvancedAdViewFactory
import com.crazecoder.flutter.mintegral.factory.SplashAdViewFactory
import com.mbridge.msdk.mbbid.out.BannerBidRequestParams
import com.mbridge.msdk.mbbid.out.SplashBidRequestParams
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

/**
 * FlutterMintegralPlugin
 */
class FlutterMintegralPlugin : FlutterPlugin, MethodCallHandler,
    ActivityAware {
    /** The MethodChannel that will the communication between Flutter and native Android
     *
     * This local reference serves to register the plugin with the Flutter Engine and unregister it
     * when the Flutter Engine is detached from the Activity */
    private var context: Context? = null
    private var activity: Activity? = null
    private var channel: MethodChannel? = null
    private var flutterPluginBinding: FlutterPluginBinding? = null
    private val sdkManager = MIntegralSDKManager()
    private var bannerAdManager: BannerAdManager? = null
    private var rewardVideoAdManager: RewardVideoAdManager? = null
    private var newInterstitialAdManager: NewInterstitialAdManager? = null


    override fun onAttachedToEngine(flutterPluginBinding: FlutterPluginBinding) {
        this.flutterPluginBinding = flutterPluginBinding
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method == "initAdSDK") {
            if (call.hasArgument("appId") && call.hasArgument("appKey")) {
                val appId = call.argument<String>("appId")
                val appKey = call.argument<String>("appKey")
                val isProtectGDPR = call.argument<Boolean>("isProtectGDPR")!!
                val isProtectCCPA = call.argument<Boolean>("isProtectCCPA")!!
                sdkManager.init(activity!!, appId, appKey, isProtectGDPR, isProtectCCPA)
            }
            result.success(null)
        } else if (call.method == "preload") {
            val adUnitId = call.argument<Any>("adUnitId").toString()
            val placementId = call.argument<Any>("placementId").toString()
            val layoutType = call.argument<Int?>("layoutType")
            sdkManager.preload(adUnitId, placementId, layoutType)
            result.success(null)
        } else if (call.method == "startSplashAd") {
            val adUnitId = call.argument<Any>("adUnitId").toString()
            val placementId = call.argument<Any>("placementId").toString()
            val launchBackgroundIdStr = call.argument<String>("launchBackgroundId")
            val isBidding = call.argument<Boolean>("isBidding")!!
            val launchBackgroundId: Int
            if (TextUtils.isEmpty(launchBackgroundIdStr)) {
                launchBackgroundId = -1
            } else {
                val res = activity!!.resources
                launchBackgroundId =
                    res.getIdentifier(launchBackgroundIdStr, "drawable", activity!!.packageName)
            }
            val intent = Intent(activity, SplashAdActivity::class.java)
            intent.putExtra("adUnitId", adUnitId)
            intent.putExtra("placementId", placementId)
            intent.putExtra("launchBackgroundId", launchBackgroundId)
            if (isBidding) {
                val params = SplashBidRequestParams(placementId, adUnitId, true, 1, 200, 200)
                sdkManager.fetchBidToken(params) { token ->
                    intent.putExtra("biddingToken", token)
                    activity!!.startActivity(intent)
                    result.success(null)
                }
            } else {
                activity!!.startActivity(intent)
                result.success(null)
            }

        } else if (call.method == "showBannerAD") {
            if (bannerAdManager == null) bannerAdManager = BannerAdManager(activity!!)
            val adUnitId = call.argument<Any>("adUnitId").toString()
            val placementId = call.argument<Any>("placementId").toString()
            val isBidding = call.argument<Boolean>("isBidding")
            val bannerSizeType = call.argument<Int>("bannerSizeType")
            val refreshTime = call.argument<Int>("refreshTime")
            if (isBidding == true) {
                val params = BannerBidRequestParams(placementId, adUnitId, 1280, 80)
                sdkManager.fetchBidToken(params) { token ->
                    bannerAdManager?.show(placementId, adUnitId, token,bannerSizeType,refreshTime = refreshTime)
                    result.success(null)
                }
                return
            }
            bannerAdManager?.show(placementId, adUnitId,bannerSizeType=bannerSizeType,refreshTime = refreshTime)
            result.success(null)
        } else if (call.method == "disposeBannerAD") {
            val adUnitId = call.argument<Any>("adUnitId").toString()
            bannerAdManager?.dispose(adUnitId)
            result.success(null)
        } else if (call.method == "showNativeAdvancedAD") {
            val adUnitId = call.argument<Any>("adUnitId").toString()
            val placementId = call.argument<Any>("placementId").toString()

            result.success(null)
        } else if (call.method == "showRewardVideoAD") {
            val adUnitId = call.argument<Any>("adUnitId").toString()
            val placementId = call.argument<Any>("placementId").toString()
            val userId = call.argument<String>("userId")
            val extraData = call.argument<String>("extraData")
            val isBidding = call.argument<Boolean>("isBidding")
            val isJustPlus = call.argument<Boolean>("isJustPlus")
            if (rewardVideoAdManager == null)
                rewardVideoAdManager = RewardVideoAdManager(activity!!, userId, extraData)
            if (isBidding == true) {
                sdkManager.fetchBidToken(placementId,adUnitId){token->
                    rewardVideoAdManager?.show(adUnitId, placementId,token,isJustPlus)
                }
            }else{
                rewardVideoAdManager?.show(adUnitId, placementId, isJustPlus = isJustPlus)
            }
            result.success(null)
        } else if (call.method == "showNewInterstitialAD") {
            val adUnitId = call.argument<Any>("adUnitId").toString()
            val placementId = call.argument<Any>("placementId").toString()
            val isBidding = call.argument<Boolean>("isBidding")
            if (newInterstitialAdManager == null)
                newInterstitialAdManager = NewInterstitialAdManager(activity!!)
            if (isBidding == true) {
                sdkManager.fetchBidToken(placementId,adUnitId){token->
                    newInterstitialAdManager?.loadAd(adUnitId, placementId,token)
                }
            }else{
                newInterstitialAdManager?.loadAd(adUnitId, placementId)
            }
            result.success(null)
        }else {
            result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
        channel!!.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        channel =
            MethodChannel(flutterPluginBinding!!.binaryMessenger, "flutter_mintegral")
        flutterPluginBinding?.platformViewRegistry?.registerViewFactory("m-banner",BannerViewFactory(activity!!))
        flutterPluginBinding?.platformViewRegistry?.registerViewFactory("m-native-advanced-ad",NativeAdvancedAdViewFactory(activity!!))
        flutterPluginBinding?.platformViewRegistry?.registerViewFactory("m-native-ad",NativeAdViewFactory(activity!!))
        flutterPluginBinding?.platformViewRegistry?.registerViewFactory("m-splash-ad",SplashAdViewFactory(flutterPluginBinding!!.binaryMessenger,activity!!))

        channel!!.setMethodCallHandler(this)
        context = flutterPluginBinding?.applicationContext
    }

    override fun onDetachedFromActivityForConfigChanges() {
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    }

    override fun onDetachedFromActivity() {
        flutterPluginBinding = null
    }

}
